pwiz_tools/Skyline/FileUI/PeptideSearch/BuildPeptideSearchLibraryControl.cs

```diff
@@ -223,28 +224,139 @@ public bool FilterForDocumentPeptides
         public string[] SearchFilenames
         {
             get { return ImportPeptideSearch.SearchFilenames; }
+            private set { SetSearchFiles(value, true); }
+        }
+
+        private void SetSearchFiles(string[] searchFiles, bool async)
+        {
+            // Set new value
+            ImportPeptideSearch.SearchFilenames = searchFiles;
+
+            // Always show sorted list of files
+            Array.Sort(ImportPeptideSearch.SearchFilenames);
+
+            var toAdd = searchFiles.ToHashSet();
 
-            private set
+            // Populate the input files list
+            gridSearchFiles.CellValueChanged -= gridSearchFiles_CellValueChanged;
+
+            var existingFiles = new HashSet<string>();
+            for (var i = gridSearchFiles.RowCount - 1; i >= 0; i--)
             {
-                // Set new value
-                ImportPeptideSearch.SearchFilenames = value;
+                var existingFile = FileCellValue.Get(gridSearchFiles.Rows[i], columnFile).File;
+                if (!toAdd.Contains(existingFile))
+                    gridSearchFiles.Rows.RemoveAt(i);
+                else
+                    existingFiles.Add(existingFile);
+            }
+            toAdd.ExceptWith(existingFiles);
 
-                // Always show sorted list of files
-                Array.Sort(ImportPeptideSearch.SearchFilenames);
+            foreach (var file in toAdd)
+            {
+                var i = gridSearchFiles.Rows.Add();
+                gridSearchFiles[columnFile.Index, i].Value = new FileCellValue(gridSearchFiles, columnFile, file);
+                gridSearchFiles[columnScoreType.Index, i].Value = new ScoreTypeCellValue();
+                gridSearchFiles[columnThreshold.Index, i] = new ThresholdCell();

```

Comment:
I think if you set the CellTemplate property of this DataGridViewTextBoxColumn to a ThresholdCell (in the BuildPeptideSearchLibraryControl constructor), that cell template would end up getting correctly cloned whenever a new row was added to the grid.